test/e2e/compact_test.go

```diff
@@ -0,0 +1,500 @@
+
+			expected: []model.Metric{
+				{
+					"a":    "1",
+					"b":    "2",
+					"ext1": "value1",
+				},
+			},
+			expectOfModBlocks: 2,
+			expectOfBlocks:    1,
+			expectOfSamples:   149,
+			expectOfSeries:    1,
+			expectOfChunks:    2,
+		},
+		{
+			name: "(full) vertically overlapping blocks with replica labels retention specified",
+			blocks: []blockDesc{
+				{
+					series:           []labels.Labels{labels.FromStrings("a", "1", "b", "2")},
+					extLset:          labels.FromStrings("ext1", "value1", "replica", "1"),
+					mint:             timestamp.FromTime(now),
+					maxt:             timestamp.FromTime(now.Add(2 * time.Hour)),
+					samplesPerSeries: 120,
+				},
+				{
+					series:           []labels.Labels{labels.FromStrings("a", "1", "b", "2")},
+					extLset:          labels.FromStrings("ext1", "value1", "replica", "2"),
+					mint:             timestamp.FromTime(now),
+					maxt:             timestamp.FromTime(now.Add(2 * time.Hour)),
+					samplesPerSeries: 120,
+				},
+				{
+					series:           []labels.Labels{labels.FromStrings("a", "1", "b", "2")},
+					extLset:          labels.FromStrings("ext1", "value1", "rule_replica", "1"),
+					mint:             timestamp.FromTime(now),
+					maxt:             timestamp.FromTime(now.Add(2 * time.Hour)),
+					samplesPerSeries: 120,
+				},
+			},
+			replicaLabels:       []string{"replica", "rule_replica"},
+			downsamplingEnabled: true,
+			retention: &retention{
+				resRaw: "0d",
+				res5m:  "5m",
+				res1h:  "5m",
+			},
+			query: "{a=\"1\"}",
+
+			expected: []model.Metric{
+				{
+					"a":    "1",
+					"b":    "2",
+					"ext1": "value1",
+				},
+			},
+			expectOfModBlocks: 3,
+			expectOfBlocks:    1,
+			expectOfSamples:   120,
+			expectOfSeries:    1,
+			expectOfChunks:    2,
+		},
+		{
+			name: "(full) vertically overlapping blocks without replica labels",
+			blocks: []blockDesc{
+				{
+					series:           []labels.Labels{labels.FromStrings("a", "1", "b", "2")},
+					extLset:          labels.FromStrings("ext1", "value1"),
+					mint:             timestamp.FromTime(now),
+					maxt:             timestamp.FromTime(now.Add(2 * time.Hour)),
+					samplesPerSeries: 2,
+				}, {
+					series:           []labels.Labels{labels.FromStrings("a", "1", "b", "2")},
+					extLset:          labels.FromStrings("ext1", "value2"),
+					mint:             timestamp.FromTime(now),
+					maxt:             timestamp.FromTime(now.Add(2 * time.Hour)),
+					samplesPerSeries: 2,
+				},
+			},
+			replicaLabels:       []string{"replica"},
+			downsamplingEnabled: true,
+			query:               "{a=\"1\"}",
+
+			expected: []model.Metric{
+				{
+					"a":    "1",
+					"b":    "2",
+					"ext1": "value1",
+				},
+				{
+					"a":    "1",
+					"b":    "2",
+					"ext1": "value2",
+				},
+			},
+			expectOfModBlocks: 0,
+			expectOfBlocks:    2,
+			expectOfSamples:   4,
+			expectOfSeries:    2,
+			expectOfChunks:    2,
+		},
+	} {
+		i := i
```

Comment:
do you really need this?

Reply:
Unfortunately yes, we're using it in a closure with `t.Run`.

Reply:
it does not really matter I think as long as it is not go routine

Reply:
Sorry about the vague answer, `t.Run` runs the passed closure in a goroutine even if you don't run your tests in parallel https://github.com/golang/go/blob/bb929b7452b5b8a6a584a94a110c98b0c6543a6e/src/testing/testing.go#L1043. I've put this here because I've seen odd results.